### ISSUE TITLE
Visual Studio Code can't find itself

### ISSUE THREAD

**Issue by GuzioMG** [microsoft/vscode#75541]
### Report Issue to prefill these. -->

- VSCode Version: 1.35.1
- OS Version: Microsoft Windows10PRO v1803 compilation17134.706 (vsCode description said: "Windows_NT x64")
- Electron: 3.1.8
- Node.JS: 10.2.0

## Steps to Reproduce:

1. Generate vsCode extension using Yeoman.
2. Open `extension.js` file.
3. Open "debugging" (the crossed bug icon) menu.
4. Make sure to BOTH check "All Exceptions" and "Uncaught Exceptions" under "Breakpoints" submenu.
5. Start debugging by pressing the "play" button on top of the "debugging" menu.
6. Wait until it crashes, saying, it can't find itself (i. e. `Cannot find module 'vscode'`).

## Does this issue occur when all extensions are disabled?:
Yes

---------------------------------------------------------------

### *Backend story*

* I was in the middle (actually closer to start) of creating a vsCode extension for Minecraft Bedrock scripting API since there are many for the `.mcfunction `language, but none of those supports the Bedrock Scripting API. And probably many of you have reached the point when you notice, that there is WAY too much code in extension.js and decide to split it between a few files. And I decided to do the same.

### *Node.JS problems*

* But when I created 3 files with copied parts of extenstion.js (including functions' definitions to provide a point of reference) and tried to refer to them by using `include('./src/fileName.js').functionName(argument, or, more, arguments); `, my JS lint, said `Property 'functionName' does not exist on type 'typeof import("c:/location/repoName/extension-id/extensionName/src/fileName")'.ts(2339)`. And then I started debugging session, to see what's going on. I don't remember the stack trace and I can't get it anymore (I'll later explain why), but I can tell, it (of course) didn't help me. So after about an hour, I figured out, I need to add `module.exports = {functionName}` on the bottom of every file (I guess, I should Google it before doing everything, not when I run out of ideas). So I did it. And JS lint errors disappeared.

### *Unexcepted error*

* But then, I decided to debug and check for errors not detected by JS lint. And there was one error. But the one I personally least excepted. I honestly didn't except such thing at all. My vsCode said, that it can't find itself. Really, I'm not kidding. Debugger threw: Error: Cannot find module 'vscode'. Also, after enabling "All Exceptions" and "Uncaught Exceptions" under "Breakpoints" submenu on "Debugging" screen (here you can also see, why did I mention it on "Steps to reproduce"), I can see this:
![image](https://user-images.githubusercontent.com/35845447/59535493-cdaa4600-8ef9-11e9-89c3-ab563fad130d.png)
and after continuing, I see this:
![image](https://user-images.githubusercontent.com/35845447/59535648-2974cf00-8efa-11e9-8932-74cdac815412.png)
The stack trace in text form:

`Debugger listening on ws://127.0.0.1:12102/19fe94d7-32a0-4e96-9bee-8422257097ea`

`For help, see: https://nodejs.org/en/docs/inspector`

`Debugger attached.`

`internal/modules/cjs/loader.js:583`

`    throw err;`

`    ^`

`  `

`Error: Cannot find module 'vscode'`

`    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)`

`    at Function.Module._load (internal/modules/cjs/loader.js:507:25)`

`    at Module.require (internal/modules/cjs/loader.js:637:17)`

`    at require (internal/modules/cjs/helpers.js:22:18)`

`    at Object.<anonymous> (c:\path\to\extension.js:1:78)`

`    at Module._compile (internal/modules/cjs/loader.js:686:14)`

`    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)`

`    at Module.load (internal/modules/cjs/loader.js:599:32)`

`    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)`

`    at Function.Module._load (internal/modules/cjs/loader.js:530:3)`

`Waiting for the debugger to disconnect...`

`Error: Cannot find module 'vscode'`

`    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)`

`    at Function.Module._load (internal/modules/cjs/loader.js:507:25)`

`    at Module.require (internal/modules/cjs/loader.js:637:17)`

`    at require (internal/modules/cjs/helpers.js:22:18)`

`    at Object.<anonymous> (c:\path\to\extension.js:1:78)`

`    at Module._compile (internal/modules/cjs/loader.js:686:14)`

`    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)`

`    at Module.load (internal/modules/cjs/loader.js:599:32)`

`    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)`

`    at Function.Module._load (internal/modules/cjs/loader.js:530:3)`

After trying to do anything (completely wiping content of "activate()", removing `module.exports = {functionName}` to get another error (I mentioned it before), creating new extension - but I'll mention it in a minute, disabling all extensions by typing `code --some-parameters-i-will-check-later` in command prompt, e. t. c),  I couldn't get rid out of this error (and here you can see, why I said:

> I don't remember the stack trace and I can't get it anymore.

No matter, what I do, I ALWAYS get `Cannot find module 'vscode'` error).

### *NOTHING can help*

* Then, I decided to start a new extension via Yeoman and manually copy-paste my changes. But before even opening any file, I decided to launch and check, did it fix anything, but my vscode said `'${file}' cannot be resolved. Please open an editor.`. So I launched `extension.js` and tried again. But nope, creating new extension also did not help. No other option left. I honestly have no idea what to do. I guess, I bricked extension development on my vscode and I can only reinstall it. Anyway, you can now see, why I said

> Generate vsCode extension using Yeoman.

on "Steps to reproduce" (because newly generated extensions also have the same problem),

> Open `extension.js` file.

on "Steps to reproduce", too (because otherwise, you'll see `'${file}' cannot be resolved. Please open an editor.`) and 

> After trying to do anything ([...] creating new extension - but I'll mention it in a minute, [...]),  I couldn't get rid out of this error

on "Unexpected error" (because I just mentioned it).

---------------------------------------------------------------

Anyway, I hope I helped you to find and fix problems with vscode by submitting this short issue. I tried to cover everything on "steps to reproduce" to make everything clear for you (I can see, that I didn't mention point 3 and 5, but I guess it's logic, you need to open "debugging" menu to configure debugger settings and start debugging [nr. 3] and you need to start it, to see what's going on [nr. 5]). I guess I'll need to reinstall vscode to un-brick extensions development. Or maybe, I'll just need to wait 1 or 2 days until someone more experienced will help me by giving probably the most logical explanation possible. Or I just helped you find an important problem with your IDE and very soon we'll all see an important update, that will fix it. I don't know, but I hope you enjoyed. See you later. (Hopefully with the solution.) Bye!

**Comment by GuzioMG** [comment]
VSCodeBot is broken...
There was (almost) about "new release"...

**Comment by Tyriar** [comment]
Did you run `npm install` in the extension directory?

**Comment by GuzioMG** [comment]
Well, I don't know, what Yeoman generator was installing there, but after setting up everything, it said 

> I'm all done. Running `npm install` for you.

**Comment by GuzioMG** [comment]
Also, as I said
> I was in THE MIDDLE of developing my extension

so I was rather debugging it earlier, right? As I said, this happened actually suddenly, after my problem with `module.exports`.

**Comment by octref** [comment]
> Make sure to BOTH check "All Exceptions" and "Uncaught Exceptions" under "Breakpoints" submenu.

Why do you need to do that? VS Code would attempt to load various modules upfront, and errors are handled.
I suggest putting a breakpoint in `activate` function. Once it's hit, you can turn on `All Exceptions`.

**Comment by GuzioMG** [comment]
Ok, I'll try doing that.

**Comment by GuzioMG** [comment]
Actually, not even try, since it simple. I should say
> OK, I'll *do* that.
Just wait a second.

**Comment by GuzioMG** [comment]
So happened EXACTLY, what I excepted - it didn't even reach this error:
![image](https://user-images.githubusercontent.com/35845447/59590092-f72ec180-90f3-11e9-89ef-711b8e7629d7.png)
 I think, now you know, why I said

> Make sure to BOTH check "All Exceptions" and "Uncaught Exceptions" under "Breakpoints" submenu.

Because if you don't do that, it will crash before entering activate. Look at an error log:

> `Error: Cannot find module 'vscode'.`

It can't `find module 'vscode'`. So the problem is with the `require()` function because this it place where all modules are coming from. However, I'm not sure why I said "BOTH" since "All Exceptions" overrides "Uncaught Exceptions" mode, so if you check only the second one, you'll get exactly same error anyway, as long as you don't catch it (which is impossible, because you can't modify `loader.js`, which is place, where the error is thrown). Also, if you put a breakpoint (as you suggested), but not in `activate` - on `const vscode = require('vscode');` line, then you'll actually prevent this error from appearing...
![image](https://user-images.githubusercontent.com/35845447/59589860-7e2f6a00-90f3-11e9-9512-d91274e3dabe.png)
...as long as you don't hit "continue":
![image](https://user-images.githubusercontent.com/35845447/59589952-a4eda080-90f3-11e9-8c24-c8cf845a02bc.png)
![image](https://user-images.githubusercontent.com/35845447/59589974-b5058000-90f3-11e9-98f0-75a463012c6c.png)

**Comment by GuzioMG** [comment]
And I'm just curious: why isn't it labeled as a "bug", since it is? While I know, why you added "needs more info", I'm not sure, why "bug" isn't added.

**Comment by octref** [comment]
![image](https://user-images.githubusercontent.com/4033249/59646761-ea20d980-91aa-11e9-81bb-159e65e63996.png)

This is wrong. You need to copy over the `.vscode` folder as well. It contains a `launch.json`, which is needed for debugging your app.

See here for an illustration. Notice it has "Run Extension" whereas you have "No configuration".
https://code.visualstudio.com/api/get-started/your-first-extension#debugging-the-extension

You can also copy from this sample: https://github.com/microsoft/vscode-extension-samples/blob/master/helloworld-sample/.vscode/launch.json

**Comment by GuzioMG** [comment]
Thank you so much! Just please please, don't close it yet. I'll check this later, since now I'm busy. And then I'll give there result and eventually close.

**Comment by GuzioMG** [comment]
It's WORKING!!!!

**Comment by GuzioMG** [comment]
## However!

**Comment by GuzioMG** [comment]
OK, it turns out, I'm simply dumb. Now I remember, that I removed that file because I thought it was autogenerated after I clicked "open `launch.json`" instead of "Cancel" button when I was on a different file than the `extension.js` script. So I removed it. And it stopped working. And then I generated a new extension. And there was no active editor, so after starting debugging, it showed "open editor" message box again. And I clicked
>"open `launch.json`" instead of "Cancel" button,

> because I thought it was autogenerated

> So I removed it.

So... Now I guess the only remaining problem is debugger not showing ( 
![image](https://user-images.githubusercontent.com/35845447/59696895-95568080-91f5-11e9-9fc5-c6d301b8259f.png)
 ) due to this error message: ![image](https://user-images.githubusercontent.com/35845447/59694968-1f9ce580-91f2-11e9-8a95-29bd2a8ecab8.png) on `"preLaunchTask": "npm: watch"` piece of code.

### Yay!

**Comment by GuzioMG** [comment]
###Well...

#Good news!!!

I decided to check out, what does the "configure tasks" button does, cause I was sure, that the solution is simple and hidden under it. And it was. Thanks to this (look at what is my mouse focused): ![image](https://user-images.githubusercontent.com/35845447/59699308-0dbf4080-91fa-11e9-8f4b-a2bf69c7887c.png). Sweet! So there is another `.json` file, which I'll need to copy from https://github.com/microsoft/vscode-extension-samples/blob/master/helloworld-sample/.vscode/! Niiiiice!

...oh. Wait. It does NOT work. Check this out (look on left, bottom corner): ![image](https://user-images.githubusercontent.com/35845447/59699938-17957380-91fb-11e9-9738-47c04301d69f.png). After clicking "Cancel" to hide this evil message of doom, and clicking "Show output" to... well... see the output, I see this: ![image](https://user-images.githubusercontent.com/35845447/59700279-b0c48a00-91fb-11e9-9497-5ab98d88d3a3.png), which look like this:

`Error: The npm task detection didn't contribute a task for the following configuration:`

`{`

`    "type": "npm",`

`    "script": "watch",`

`    "problemMatcher": "$tsc-watch",`

`    "isBackground": true,`

`    "presentation": {`

`        "reveal": "never"`

`    },`

`    "group": {`

`        "kind": "build",`

`        "isDefault": true`

`    }`

`}`

`The task will be ignored.`

in text form.

**Comment by GuzioMG** [comment]
# UPDATE!!!! VScode in DEV mode starts, but I just noticed, that my extension is not detected.

**Comment by octref** [comment]
> I just noticed, that my extension is not detected.

Before starting everything from scratch, can you read through the beginner sections? You need to run the "Hello World" command in the newly opened window to activate it.

- https://code.visualstudio.com/api/get-started/your-first-extension (this page even has a video of it)
- https://code.visualstudio.com/api/get-started/extension-anatomy

> "preLaunchTask": "npm: watch"

That corresponds to this line: https://github.com/microsoft/vscode-extension-samples/blob/master/helloworld-sample/package.json#L30

**Comment by octref** [comment]
Btw I closed because my last reply should answer all your questions. Feel free to reopen if that's not the case.

**Comment by GuzioMG** [comment]
# Hell yeah!!!

OK. I was idle for a while, but now I decided to check one thing. At this point, I need to reveal you my repo link. It's github.com/GuzioMg/McIDE.js. But why I did it? Well, if you open up that repo, you're not going to see vscode extension folder structure. You're going to see the `mcide-4jsapi` folder instead. Why? Because I generated my extension there. And  `mcide-4jsapi` is my extension ID. The 4jsapi stands there, to inform that this is 4 ([OR:] for) the JS (Bedrock Scripting) API, to avoid mistaking my extension with [mcide](https://marketplace.visualstudio.com/items?itemName=intsuc.mcide), which is targeted to `.mcfunction` (a.k.a. Minecode) language instead. And inside that folder, there is another one, called `mcide`. It was autogenerated by Yeoman because I typed "McIDE.js" and `.js` was cut and the entire phase was lowercased. Not sure why, though. Anyway, if you open that folder, you're going to see the normal structure of VSC extension. _INCLUDING `.vscode` folder. And `node-modules`._ And everything else required for this extension to work

---------------------------------------------
![image](https://user-images.githubusercontent.com/35845447/59981845-2ff9f980-9612-11e9-84c3-d742f5a3d56f.png)
---------------------------------------------
`Booting-up "McIDE.js"...`
`Registring commands...`
`Registring command [1/3] - "Sync MC events in this file" (ID: "extension.syncmceventregistry")...`
`Registred  command [1/3] - "Sync MC events in this file" (ID: "extension.syncmceventregistry").`
`Registering command [2/3] - "Autogenerate new MC event code here" (ID: "extension.createminecrafteventcode")...`
`Registered  command [2/3] - "Autogenerate new MC event code here" (ID: "extension.createminecrafteventcode").`
`Registering command [3/3] - "McIDE.js" (ID: "extension.mcide-4jsapi")...`
`Registered  command [3/3] - "McIDE.js" (ID: "extension.mcide-4jsapi").`
`Registered commands. 0 errors caught. Initialization took 0 ms.`
`"McIDE.js" fully loaded successfully! 0 errors caught. Initialization took 0 ms.`
---------------------------------------------
![bitmoji](https://render.bitstrips.com/v2/cpanel/430025ee-b231-4e76-b651-9909984b7b68-adafaadc-4555-4b2b-807d-942032e2e7cc-v1.png?transparent=1&palette=1&width=246)

**Comment by GuzioMG** [comment]
SO, everything I can say you, @octref, is simple "Thank you!".

> It turns out, I'm simply dumb. 

EVERYTHING, I should do was simply open VSC workspace on another folder. I'm not sure, why I say "thank you" since (after years of trying) I figured it out myself. But thank you for your attention. And time. And totally 20 comments on this issue, which it took us to figure it out. I guess, now you can remove the "*question" label since this wasn't question. It was my mistake. This comment is actually a filler, just to reach 20 comments because I could say "Thak you" on the previous one. But I did not.

Thank you!

Hopefully, see you soon. And hopefully not on another issue, but during verifying McIDE.js (if extensions are verified).

Goodbye!!!!!

-Guzio

![bitmoji](https://render.bitstrips.com/v2/cpanel/9d9f5e18-ccc2-49a8-92e5-5196b1caad81-adafaadc-4555-4b2b-807d-942032e2e7cc-v1.png?transparent=1&palette=1&width=246)

---------------------------------------------
![image](https://user-images.githubusercontent.com/35845447/59981845-2ff9f980-9612-11e9-84c3-d742f5a3d56f.png)
---------------------------------------------
`Booting-up "McIDE.js"...`
`Registring commands...`
`Registring command [1/3] - "Sync MC events in this file" (ID: "extension.syncmceventregistry")...`
`Registred  command [1/3] - "Sync MC events in this file" (ID: "extension.syncmceventregistry").`
`Registering command [2/3] - "Autogenerate new MC event code here" (ID: "extension.createminecrafteventcode")...`
`Registered  command [2/3] - "Autogenerate new MC event code here" (ID: "extension.createminecrafteventcode").`
`Registering command [3/3] - "McIDE.js" (ID: "extension.mcide-4jsapi")...`
`Registered  command [3/3] - "McIDE.js" (ID: "extension.mcide-4jsapi").`
`Registered commands. 0 errors caught. Initialization took 0 ms.`
`"McIDE.js" fully loaded successfully! 0 errors caught. Initialization took 0 ms. Or 10 days. If you consider date of issue creation.`
---------------------------------------------
![bitmoji](https://render.bitstrips.com/v2/cpanel/430025ee-b231-4e76-b651-9909984b7b68-adafaadc-4555-4b2b-807d-942032e2e7cc-v1.png?transparent=1&palette=1&width=246)
